File: resources/js/utils/bugReporter.js
/**
 * TPIX TRADE — ตัวรายงานบั๊กของหน้าเว็บ → ระบบกลาง xman studio (ผ่าน relay ของเรา)
 *
 * เจ้าของสั่ง: "ทำระบบ bug report … เพื่อให้ตรวจสอบได้ทันที ไม่เดา"
 *  - ดัก error ที่ไม่มีใครจับ (window.error / unhandledrejection / Vue errorHandler)
 *  - breadcrumb: การนำทาง + คำขอ API ที่ล้มเหลว 40 รายการล่าสุด แนบทุกรายงาน
 *  - ล้างความลับก่อนส่ง · กันซ้ำ 1 ชั่วโมงต่อรอยเดียวกัน · ไม่เกิน 5 รายงาน/นาที
 *
 * Developed by Xman Studio
 */
import axios from 'axios';
import { router } from '@inertiajs/vue3';

const ENDPOINT = '/api/v1/bug-reports';
const PRODUCT = 'tpix-web';
const MAX_CRUMBS = 40;
const DEDUPE_MS = 60 * 60 * 1000;
const PER_MINUTE = 5;

const crumbs = [];
const recent = new Map();
let minuteStart = Date.now();
let sentThisMinute = 0;

const version = () => document.querySelector('meta[name="app-version"]')?.content || 'web';

function deviceId() {
    try {
        let id = localStorage.getItem('tpix_web_device_id');
        if (!id) {
            id = Array.from(crypto.getRandomValues(new Uint8Array(16))).map(b => b.toString(16).padStart(2, '0')).join('');
            localStorage.setItem('tpix_web_device_id', id);
        }
        return id;
    } catch (_) {
        return 'unknown';
    }
}

const HEX64 = /(?<![0-9a-fA-F])(0x)?[0-9a-fA-F]{64}(?![0-9a-fA-F])/g;
const SIG = /0x[0-9a-fA-F]{130}/g;
const BEARER = /(Bearer\s+)[A-Za-z0-9._-]{8,}/gi;
const MNEMONIC = /\b(?:[a-z]{3,8}\s+){11,23}[a-z]{3,8}\b/g;

function scrub(text) {
    return String(text ?? '')
        .replace(SIG, m => `${m.slice(0, 10)}…[sig]`)
        .replace(HEX64, '[secret-64hex]')
        .replace(BEARER, '$1[token]')
        .replace(MNEMONIC, '[mnemonic?]');
}

export function breadcrumb(text) {
    const line = `${new Date().toISOString().slice(11, 19)} ${scrub(text)}`;
    crumbs.push(line.length > 200 ? `${line.slice(0, 200)}…` : line);
    if (crumbs.length > MAX_CRUMBS) crumbs.shift();
}

export function reportBug({ title, description, type = 'bug', severity = 'moderate', priority = 'medium', metadata = {}, stack = null, dedupe = true }) {
    const cleanTitle = scrub(title).slice(0, 250);
    const cleanStack = stack ? scrub(stack).slice(0, 12000) : null;
    const fingerprint = `${type}|${cleanTitle}|${cleanStack ? cleanStack.split('\n')[0] : ''}`;

    if (dedupe) {
        const last = recent.get(fingerprint);
        if (last && Date.now() - last < DEDUPE_MS) return false;
    }
    recent.set(fingerprint, Date.now());

    if (Date.now() - minuteStart > 60_000) { minuteStart = Date.now(); sentThisMinute = 0; }
    if (++sentThisMinute > PER_MINUTE) return false;

    const body = {
        product_name: PRODUCT,
        product_version: version(),
        app_version: version().slice(0, 20),
        os_version: navigator.userAgent.slice(0, 100),
        device_id: deviceId(),
        report_type: type,
        title: cleanTitle,
        description: scrub(description).slice(0, 20000),
        stack_trace: cleanStack,
        priority,
        severity,
        metadata: {
            url: location.href,
            viewport: `${window.innerWidth}x${window.innerHeight}`,
            language: navigator.language,
            breadcrumbs: [...crumbs],
            reported_at: new Date().toISOString(),
            ...metadata,
        },
    };

    try {
        // keepalive: ส่งให้ทันแม้หน้ากำลังปิด
        fetch(ENDPOINT, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
            body: JSON.stringify(body),
            keepalive: true,
        }).catch(() => {});
    } catch (_) { /* ไม่มีอะไรให้ทำ — รายงานบั๊กห้ามทำให้หน้าพังซ้ำ */ }

    return true;
}

/** ติดตั้งครั้งเดียวใน app.js หลังสร้าง Vue app */
export function installBugReporter(app) {
    breadcrumb(`page load ${location.pathname}`);

    window.addEventListener('error', (event) => {
        // error จากสคริปต์ข้ามโดเมนไม่มีรายละเอียด (Script error.) รายงานไปก็อ่านไม่ออก
        if (!event.message || event.message === 'Script error.') return;
        reportBug({
            title: event.message,
            description: `${event.message}\nat ${event.filename || '?'}:${event.lineno || 0}:${event.colno || 0}`,
            type: 'crash',
            severity: 'major',
            stack: event.error?.stack,
        });
    });

    window.addEventListener('unhandledrejection', (event) => {
        const reason = event.reason;
        const message = reason?.message || String(reason);
        // axios ยกเลิก/หมดเวลา ไม่ใช่บั๊ก
        if (reason?.code === 'ERR_CANCELED') return;
        reportBug({
            title: `unhandled: ${message}`,
            description: message,
            type: 'crash',
            severity: 'moderate',
            stack: reason?.stack,
            metadata: { status: reason?.response?.status, url: reason?.config?.url },
        });
    });

    app.config.errorHandler = (err, instance, info) => {
        console.error('[TPIX] Vue Error:', err, info);
        reportBug({
            title: `vue: ${err?.message || String(err)}`,
            description: `${err?.message || String(err)}\ninfo: ${info}`,
            type: 'crash',
            severity: 'major',
            stack: err?.stack,
            metadata: { component: instance?.$options?.name || instance?.$?.type?.__name || null, info },
        });
    };

    router.on('navigate', (event) => breadcrumb(`navigate ${event.detail?.page?.url || ''}`));

    axios.interceptors.response.use(
        (response) => response,
        (error) => {
            const status = error?.response?.status;
            const url = error?.config?.url || '';
            if (status && url) breadcrumb(`api ${(error.config?.method || 'get').toUpperCase()} ${url} → ${status}`);
            return Promise.reject(error);
        },
    );
}

File: mobile-flutter/lib/services/bug_reporter.dart
/// TPIX — ตัวรายงานบั๊กไปยังระบบกลางของ xman studio
///
/// เจ้าของสั่ง: "ทำระบบ bug report หลังบ้านไว้สำหรับรายงานข้อผิดพลาดจากแอพและโปรแกรม
/// ทั้งหมด เพื่อให้ตรวจสอบได้ทันที ไม่เดา"
///
/// ส่งเข้า https://xman4289.com/api/v1/bug-reports (ระบบกลางที่ทุกแอปในบ้านใช้อยู่แล้ว
/// มีหน้าแอดมิน สถิติ และอ่านผ่าน GET สาธารณะได้) — ไม่สร้างระบบซ้อน
///
/// สามอย่างที่ทำ:
///  1. จับ error ที่ไม่มีใครดัก (FlutterError / PlatformDispatcher / zone) → รายงาน crash
///  2. breadcrumb — บันทึกเหตุการณ์สำคัญ 40 รายการล่าสุด (deep link, เซ็น, ยืนยัน, API)
///     แนบไปกับทุกรายงาน → คนอ่านเห็นว่า "ก่อนพังเกิดอะไรขึ้นบ้าง" ไม่ต้องเดา
///  3. ผู้ใช้กด "รายงานปัญหา" เอง → ส่งข้อความ + สภาพแอปตอนนั้น + breadcrumb
///
/// ความปลอดภัย: ล้างข้อมูลลับก่อนส่งเสมอ (กุญแจ 64 hex, ลายเซ็น, mnemonic, โทเคน)
/// คิวออฟไลน์ใน SharedPreferences · กันซ้ำด้วย fingerprint 1 ชั่วโมง · ไม่บล็อก UI
///
/// Developed by Xman Studio
library;

import 'dart:async';
import 'dart:convert';
import 'dart:io' show Platform;
import 'dart:math';
import 'dart:ui' show PlatformDispatcher;

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:package_info_plus/package_info_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';

class BugReporter {
  BugReporter._();
  static final BugReporter I = BugReporter._();

  static const _endpoint = 'https://xman4289.com/api/v1/bug-reports';
  static const _queueKey = 'bug_reports_queue_v1';
  static const _deviceKey = 'bug_reports_device_id_v1';
  static const _maxQueue = 30;
  static const _maxCrumbs = 40;
  static const _dedupeWindow = Duration(hours: 1);
  static const _timeout = Duration(seconds: 10);

  String product = 'tpix-app';
  String version = '0';
  String build = '0';
  String? deviceId;

  bool _installed = false;
  bool _flushing = false;
  final List<String> _crumbs = [];
  final Map<String, DateTime> _recent = {};
  final List<Map<String, dynamic>> _queue = [];

  /// สภาพแอปตอนรายงาน — ผู้เรียกตั้งให้ (เช่น ชนิดกระเป๋า ยืนยันแล้วไหม เชน) ห้ามใส่ความลับ
  Map<String, dynamic> Function()? snapshot;

  bool get installed => _installed;

  /// เรียกครั้งเดียวใน main() ก่อน runApp — ติดตั้งตัวดัก error ทั้งสามชั้น
  Future<void> install({required String product}) async {
    if (_installed) return;
    _installed = true;
    this.product = product;

    try {
      final info = await PackageInfo.fromPlatform();
      version = info.version;
      build = info.buildNumber;
    } catch (_) {}

    try {
      final prefs = await SharedPreferences.getInstance();
      deviceId = prefs.getString(_deviceKey);
      if (deviceId == null || deviceId!.isEmpty) {
        // เลขสุ่มของการติดตั้ง — พอจับคู่รายงานจากเครื่องเดียวกัน ไม่ผูกกับตัวเครื่องจริง
        deviceId = _randomId();
        await prefs.setString(_deviceKey, deviceId!);
      }
      final raw = prefs.getString(_queueKey);
      if (raw != null && raw.isNotEmpty) {
        final decoded = jsonDecode(raw);
        if (decoded is List) {
          _queue.addAll(decoded.whereType<Map>().map((m) => Map<String, dynamic>.from(m)));
        }
      }
    } catch (_) {}

    final previous = FlutterError.onError;
    FlutterError.onError = (details) {
      previous?.call(details);
      unawaited(reportError(
        details.exception,
        details.stack,
        context: 'FlutterError ${details.library ?? ''}'.trim(),
      ));
    };

    final previousDispatcher = PlatformDispatcher.instance.onError;
    PlatformDispatcher.instance.onError = (error, stack) {
      unawaited(reportError(error, stack, context: 'PlatformDispatcher', fatal: true));
      return previousDispatcher?.call(error, stack) ?? true;
    };

    breadcrumb('app start $product v$version+$build ${_os()}');
    unawaited(flush());
  }

  /// บันทึกเหตุการณ์สำคัญ — สั้นๆ ไม่มีความลับ (ถูกล้างอีกชั้นตอนส่งอยู่ดี)
  void breadcrumb(String text) {
    final line = '${DateTime.now().toIso8601String().substring(11, 19)} ${_scrub(text)}';
    _crumbs.add(line.length > 200 ? '${line.substring(0, 200)}…' : line);
    if (_crumbs.length > _maxCrumbs) _crumbs.removeAt(0);
    if (kDebugMode) debugPrint('[crumb] $line');
  }

  List<String> get breadcrumbs => List.unmodifiable(_crumbs);

  /// error ที่ดักได้เอง (try/catch) หรือจากตัวดักอัตโนมัติ
  Future<void> reportError(
    Object error,
    StackTrace? stack, {
    String? context,
    bool fatal = false,
  }) {
    final message = _scrub(error.toString());
    return report(
      title: '${context != null ? '[$context] ' : ''}${_firstLine(message)}',
      description: message,
      type: 'crash',
      severity: fatal ? 'critical' : 'major',
      priority: fatal ? 'high' : 'medium',
      stack: stack?.toString(),
      metadata: {'context': context, 'fatal': fatal},
    );
  }

  /// รายงานทั่วไป — ใช้ทั้งจากโค้ด (เหตุการณ์ที่รู้ว่าผิดปกติ) และจากปุ่มของผู้ใช้
  Future<void> report({
    required String title,
    required String description,
    String type = 'bug',
    String severity = 'moderate',
    String priority = 'medium',
    Map<String, dynamic>? metadata,
    String? stack,
    bool dedupe = true,
  }) async {
    if (!_installed) return;

    final cleanTitle = _clip(_scrub(title), 250);
    final cleanDescription = _clip(_scrub(description), 20000);
    final cleanStack = stack == null ? null : _clip(_scrub(stack), 12000);

    final fingerprint = '$type|$cleanTitle|${cleanStack != null ? _firstLine(cleanStack) : ''}';
    if (dedupe) {
      final last = _recent[fingerprint];
      if (last != null && DateTime.now().difference(last) < _dedupeWindow) return;
    }
    _recent[fingerprint] = DateTime.now();

    Map<String, dynamic> state = const {};
    try {
      state = snapshot?.call() ?? const {};
    } catch (_) {}

    final body = <String, dynamic>{
      'product_name': product,
      'product_version': version,
      'app_version': _clip(version, 20),
      'os_version': _clip(_os(), 100),
      'device_id': deviceId,
      'report_type': type,
      'title': cleanTitle,
      'description': cleanStack == null
          ? cleanDescription
          : '$cleanDescription\n\n--- stack ---\n$cleanStack',
      'stack_trace': cleanStack,
      'priority': priority,
      'severity': severity,
      'metadata': {
        'build': build,
        'platform': Platform.operatingSystem,
        'state': _scrubMap(state),
        'breadcrumbs': List<String>.from(_crumbs),
        'reported_at': DateTime.now().toIso8601String(),
        ..._scrubMap(metadata ?? const {}),
      },
    };

    _queue.add(body);
    if (_queue.length > _maxQueue) _queue.removeAt(0);
    await _persist();
    unawaited(flush());
  }

  /// ส่งคิวที่ค้าง — ทีละรายการ ล้มเหลวเก็บไว้ส่งรอบหน้า (ไม่ทิ้ง ไม่บล็อก)
  Future<void> flush() async {
    if (_flushing || _queue.isEmpty) return;
    _flushing = true;
    try {
      while (_queue.isNotEmpty) {
        final item = _queue.first;
        final ok = await _send(item);
        if (!ok) break;
        _queue.removeAt(0);
        await _persist();
      }
    } finally {
      _flushing = false;
    }
  }

  Future<bool> _send(Map<String, dynamic> body) async {
    try {
      final res = await http
          .post(
            Uri.parse(_endpoint),
            headers: {
              'Content-Type': 'application/json',
              'Accept': 'application/json',
              // WAF ของเซิร์ฟเวอร์กลางตอบ 403 ให้ User-Agent ตั้งต้นของไลบรารี
              'User-Agent': '$product/$version ($build)',
            },
            body: jsonEncode(body),
          )
          .timeout(_timeout);
      if (res.statusCode == 201 || res.statusCode == 200) return true;
      // 422 = รูปแบบผิด ส่งซ้ำก็ไม่ผ่าน ทิ้งเพื่อไม่ให้คิวตัน
      if (res.statusCode == 422) {
        debugPrint('BugReporter: rejected ${res.body.length > 200 ? res.body.substring(0, 200) : res.body}');
        return true;
      }
      return false;
    } catch (e) {
      debugPrint('BugReporter.send: ${e.runtimeType}');
      return false;
    }
  }

  Future<void> _persist() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      if (_queue.isEmpty) {
        await prefs.remove(_queueKey);
      } else {
        await prefs.setString(_queueKey, jsonEncode(_queue));
      }
    } catch (_) {}
  }

  // ── ล้างความลับ ─────────────────────────────────────────────────────

  static final _hex64 = RegExp(r'(?<![0-9a-fA-F])(0x)?[0-9a-fA-F]{64}(?![0-9a-fA-F])');
  static final _hexSig = RegExp(r'0x[0-9a-fA-F]{130}');
  static final _bearer = RegExp(r'(Bearer\s+)[A-Za-z0-9._\-]{8,}', caseSensitive: false);
  static final _sessionHeader = RegExp(r'(X-Wallet-Session[":\s]+)[0-9a-f]{64}', caseSensitive: false);
  // 12+ คำภาษาอังกฤษตัวเล็กติดกัน = น่าจะเป็น mnemonic ไม่เสี่ยงปล่อยผ่าน
  static final _mnemonic = RegExp(r'\b(?:[a-z]{3,8}\s+){11,23}[a-z]{3,8}\b');

  static String _scrub(String input) {
    var out = input;
    out = out.replaceAllMapped(_hexSig, (m) => '${m.group(0)!.substring(0, 10)}…[sig]');
    out = out.replaceAll(_hex64, '[secret-64hex]');
    out = out.replaceAllMapped(_bearer, (m) => '${m.group(1)}[token]');
    out = out.replaceAllMapped(_sessionHeader, (m) => '${m.group(1)}[token]');
    out = out.replaceAll(_mnemonic, '[mnemonic?]');
    return out;
  }

  static Map<String, dynamic> _scrubMap(Map<String, dynamic> map) {
    final out = <String, dynamic>{};
    map.forEach((k, v) {
      final key = k.toLowerCase();
      if (key.contains('mnemonic') || key.contains('private') || key.contains('secret') || key.contains('pin')) {
        out[k] = '[redacted]';
      } else if (v is String) {
        out[k] = _scrub(v);
      } else if (v is Map) {
        out[k] = _scrubMap(Map<String, dynamic>.from(v));
      } else if (v is List) {
        out[k] = v.map((e) => e is String ? _scrub(e) : e).toList();
      } else {
        out[k] = v;
      }
    });
    return out;
  }

  static String _firstLine(String s) {
    final i = s.indexOf('\n');
    return (i == -1 ? s : s.substring(0, i)).trim();
  }

  static String _clip(String s, int max) => s.length <= max ? s : '${s.substring(0, max)}…';

  static String _os() {
    try {
      return '${Platform.operatingSystem} ${Platform.operatingSystemVersion}';
    } catch (_) {
      return 'unknown';
    }
  }

  static String _randomId() {
    final rng = Random.secure();
    return List<int>.generate(16, (_) => rng.nextInt(256))
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
  }
}
